Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachineManager.cs

```diff
@@ -338,7 +338,7 @@ public static async Task<VirtualMachine> ConnectInternalAsync (Socket dbg_sock,
 
 		public static void CancelConnection (IAsyncResult asyncResult)
 		{
-			((Socket)asyncResult.AsyncState).Close ();
+			((Socket)asyncResult.AsyncState)?.Close ();
```

Comment:
What sets the asyncstate to null?

Reply:
It happens at some point in the same `VirtualMachineManager` class because the method that returns the `IAsyncResult` (BeginListen) doesn't end up correctly.

Here it starts listening for connections: https://github.com/mono/debugger-libs/blob/main/Mono.Debugging.Soft/SoftDebuggerSession.cs#L247

Here it calls an internal listen method: https://github.com/mono/debugger-libs/blob/cdf25369c56e6006f64b63fd3ea4d817c2cf8da8/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachineManager.cs#L240

Here it ends up abnormally: https://github.com/mono/debugger-libs/blob/cdf25369c56e6006f64b63fd3ea4d817c2cf8da8/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachineManager.cs#L183

Then, we are catching an exception on the XamarinVS side and calling the `EndSession` method of the debugger session, which at that point has the `IAsyncResult.AsyncState` as null (presumably because it never completed correctly the `BeginListen` method).

Reply:
Thanks for the clarification. Can please we add that info in code comments comments as well? :D

Reply:
Done